cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -467,6 +467,20 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) errors.AutoscalerError
 			return errors.ToAutoscalerError(errors.InternalError, err)
 		}
 	}
+	// Nodes that are registered but not yet ready are classified as NotStarted by CSR. Above, we inject "upcoming" replacement
+	// nodes for them that are faked to appear ready, so that we can pack unschedulable pods on them and not trigger another scale-up.
+	// The initial nodes have to be filtered out of the all nodes list so that scale-down can't consider them as candidates. Otherwise, with
+	// aggressive scale-down settings, we could be removing the nodes before they have a chance to first become ready (the duration
+	// of which should be unrelated to the scale-down settings).
+	allNodes = subtractNodesByName(allNodes, a.clusterStateRegistry.GetClusterReadiness().NotStarted)
```

Comment:
What you want to filter here is in my opinion upcoming nodes (since those are the nodes that would have templateNode added for them). I think that by having 2 different paths for providing the same information we're setting ourselves up for future bugs when those get out of sync.

Also - I don't think we should expose Readiness outside of CSR. I know it's technically public today, but so far it was only for the purpose of pushing it to metrics. My main reason is that readiness is a huge mess, where some values are additive and some are not (ResourceUnready is a subset of Unready, but other stuff can be added etc) and we keep changing semantics completely when we're fixing various bugs.

As an example - we used to have LongNotStarted for a long time, that would change the logic here.

Reply:
Upcoming as in "fake replacements" are not added to `allNodes` in the first place. Upcoming as in "the nodes that are being replaced" consist of Unregistered (for which there are no names so nothing to filter), and NotStarted - which have names and should be filtered out of `allNodes`. So it's only "registered upcoming" nodes that should be filtered out.

Good points about having one path and not leaking the details of CSR. I changed CSR so that it is the one determining which nodes should be filtered without leaking any details (apart from the lists being public, but we've had that with the ints as well), in the same method that determines upcoming node counts.